### PR TITLE
Refactors radix cache for extra key support

### PR DESCRIPTION
**Pull request by JustinTong0323**
## Motivation
In specific scenarios, it is necessary to isolate the kvcache for certain requests. For instance, requests utilizing different LoRA adapters, or requests for which we do not desire sharing the kvcache with others(ref https://github.com/sgl-project/sglang/issues/9163).

This PR refactors the Key in the Radix cache by adding an `extra_key` field. Only requests/tokens in the cache that have the same `extra_key` can share the prefix KV cache.

I would implement the logic for cache salting in following PR.

## Modifications

This commit refactors the radix cache to support an extra key (e.g., lora_id, cache_salt) for classifying requests.

The changes include:

- Adds `extra_key` to the `Req` class 
- Replace raw token_ids key to `BaseKey` class  to facilitate request classification.
- Modifies the `match_prefix` method in `RadixCache` and `SchedulePolicy` to incorporate the `BaseKey` when matching prefixes.
- **Removes the LoRARadixCache**, as the base RadixCache can now handle LoRA IDs through the `extra_key` mechanism.

## TODO List:
- [x] Add support for HiRadixCache
- [x] Support SWA radix cache
- [x] Compatible for cpp/lmc
- [ ] Support radix cache cpp for extra keys (need modify cpp codes, in following PR
- [ ] others I may not considered

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: test/srt/test_swa_unittest.py [hunk 1]
```diff
@@ -4,7 +4,8 @@
 
 from sglang.srt.mem_cache.allocator import SWAKVPool, SWATokenToKVPoolAllocator
 from sglang.srt.mem_cache.memory_pool import ReqToTokenPool
-from sglang.srt.mem_cache.radix_cache import SWARadixCache
+from sglang.srt.mem_cache.radix_cache import BaseKey
```
Comment (ispobock) [inline test/srt/test_swa_unittest.py]: Do we have unit test for radix cache? Maybe we can create one.

File: python/sglang/srt/mem_cache/radix_cache.py [hunk 2]
```diff
@@ -40,14 +40,38 @@
     from sglang.srt.managers.schedule_batch import Req
 
 
+class BaseKey:
```
Comment (ispobock) [inline python/sglang/srt/mem_cache/radix_cache.py]: qq: Why call it base key? How about putting it to base_prefix_cache?

File: test/srt/test_swa_unittest.py (hunk 1 above)
Comment (JustinTong0323) [reply test/srt/test_swa_unittest.py]: We have `sglang/test/srt/test_radix_attention.py`

File: python/sglang/srt/mem_cache/radix_cache.py (hunk 2 above)
Comment (JustinTong0323) [reply python/sglang/srt/mem_cache/radix_cache.py]: I think there could be additional Key schemas inheriting from this base class in the future. These schemas could contain more information, like for hicache, we could attach data such as location (I'm not familiar with hicache, this is just a random thought).

Comment (JustinTong0323) [reply python/sglang/srt/mem_cache/radix_cache.py]: Also I noticed BaseKey is not used in chunked cache, so I suppose it should placed in radix cache.

File: test/srt/test_swa_unittest.py (hunk 1 above)
Comment (ispobock) [reply test/srt/test_swa_unittest.py]: I see. It seems an e2e test, not unit test. Unit test will test the basic functions like insert/match/eviction.

File: python/sglang/srt/mem_cache/radix_cache.py (hunk 2 above)
Comment (ispobock) [reply python/sglang/srt/mem_cache/radix_cache.py]: @xiezhq-hermann what do you think?

File: python/sglang/srt/mem_cache/radix_cache.py [hunk 3]
```diff
@@ -192,12 +236,12 @@ def match_prefix(self, key: List[int], **kwargs) -> MatchResult:
...
         )
 
-    def insert(self, key: List, value=None, chunked=False):
+    def insert(self, key: BaseKey, value=None, chunked=False):
         if self.disable:
             return 0
 
         if value is None:
-            value = [x for x in key]
+            value = torch.tensor([x for x in key.token_ids], dtype=torch.int64)
```
Comment (JustinTong0323) [inline python/sglang/srt/mem_cache/radix_cache.py]: Note:  this line is to fix a minor error, if value is `list`, an error would be raised when execute `torch.cat(value)`

File: test/srt/test_swa_unittest.py (hunk 1 above)
Comment (JustinTong0323) [reply test/srt/test_swa_unittest.py]: Added.

File: python/sglang/srt/mem_cache/radix_cache.py (hunk 3 above)
Comment (JustinTong0323) [reply python/sglang/srt/mem_cache/radix_cache.py]: but it seems not used... not sure if we could directly delete this `if value is None:`

File: python/sglang/srt/mem_cache/radix_cache.py [hunk 4]
```diff
@@ -137,29 +181,29 @@ def __init__(
...
         self.protected_size_ = 0
         self._record_all_cleared_event()
 
-    def match_prefix(self, key: List[int], **kwargs) -> MatchResult:
+    def match_prefix(self, key: BaseKey, **kwargs) -> MatchResult:
         """Find the matching prefix from the radix tree.
         Args:
-            key: A list of token IDs to find a matching prefix.
         Returns:
+            key: A BaseKey contains a list of token IDs to find a matching prefix.
```
Comment (Fridge003) [inline python/sglang/srt/mem_cache/radix_cache.py]: Here we need some explanation on extra_key

**Comment by Fridge003** [comment]
@JustinTong0323 Please fix the conflict

**Comment by Fridge003** [comment]
Seems really clean and tidy to me.
I feel the key of this PR is to make sure it doesn't break all kinds of features (hi-cache, swa-radix...) So please make sure this PR isn't merged until all the related tests can pass

File: python/sglang/srt/mem_cache/radix_cache.py [hunk 5]
```diff
@@ -203,12 +274,12 @@ def match_prefix(self, key: List[int], **kwargs) -> MatchResult:
...
         )
 
-    def insert(self, key: List, value=None, chunked=False):
+    def insert(self, key: BaseKey, value=None, chunked=False):
         if self.disable:
             return 0
 
         if value is None:
-            value = [x for x in key]
+            value = torch.tensor(key.token_ids, dtype=torch.int64)
```
Comment (JustinTong0323) [inline python/sglang/srt/mem_cache/radix_cache.py:282]: Note: this line is to fix a minor error, if value is list, an error would be raised when execute` torch.cat(value) `
But it seems not used... not sure if we could directly delete this `if value is None:`

**Review by Fridge003 — APPROVED** [review]
LGTM

File: python/sglang/srt/mem_cache/radix_cache.py (hunk 2 above)
Comment (xiezhq-hermann) [reply python/sglang/srt/mem_cache/radix_cache.py]: I think called it as radix_key might be more straightforward, it is specific to radix trees and can contain different information, including the token ids and the extra information/

**Review by xiezhq-hermann — APPROVED** [review]
Looks good to me, would suggest to rename `BaseKey` to `RadixKey` though

File: python/sglang/srt/mem_cache/radix_cache.py (hunk 2 above)
Comment (hnyls2002) [reply python/sglang/srt/mem_cache/radix_cache.py]: @JustinTong0323 Do not call this class `BaseKey`. Name it `RadixKey` or others.

File: python/sglang/srt/mem_cache/radix_cache.py (hunk 5 above)
Comment (hnyls2002) [reply python/sglang/srt/mem_cache/radix_cache.py:282]: This value will not be None unless in the `__main__` function in this file. Only used for testing, I think.

**Review by hnyls2002 — APPROVED** [review]
